### PR TITLE
Add deal id and rewardred inventory logic

### PR DESCRIPTION
**Pull request by samiul-shanto**
- Handle deal id
- Handle Rewarded inventory

### CODE REVIEW

File: adapters/improvedigital/improvedigital.go [hunk 1]
```diff
@@ -36,6 +47,15 @@ func (a *ImprovedigitalAdapter) MakeRequests(request *openrtb2.BidRequest, reqIn
 }
 
 func (a *ImprovedigitalAdapter) makeRequest(request openrtb2.BidRequest, imp openrtb2.Imp) (*adapters.RequestData, error) {
+	// Handle Rewarded Inventory
+	extImp, err := getImpExtWithRewardedInventory(imp)
+	if err != nil {
+		return nil, err
+	}
+	if len(extImp) > 0 {
+		imp.Ext = extImp
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigital.go:63]: Please set just `is_rewarded_inventory` as this will overwrite entire imp.ext object with any properties that could be there already.

Comment (samiul-shanto) [reply adapters/improvedigital/improvedigital.go:63]: `getImpExtWithRewardedInventory` method keep all properties of imp.ext object and only add `is_rewarded_inventory` object when data found at prebid.is_rewarded_inventory.

We also did this way for "Google Additional consent" (https://github.com/ImproveDigital/prebid-server/blob/dealid-rewarded-inventory/adapters/improvedigital/improvedigital.go#L68) based on reviewer feedback.

File: adapters/improvedigital/improvedigital.go [hunk 2]
```diff
@@ -13,10 +14,21 @@ import (
 	"github.com/prebid/prebid-server/openrtb_ext"
 )
 
+const BuyingTypeRTB = "rtb"
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigital.go]: should be `buyingTypeRTB`

File: adapters/improvedigital/improvedigital.go [hunk 3]
```diff
@@ -222,3 +256,36 @@ func (a *ImprovedigitalAdapter) getAdditionalConsentProvidersUserExt(request ope
...
+	rewardedInventory, foundRewardedInventory := prebidMap[isRewardedInventory]
+	if !foundRewardedInventory {
+		return nil, nil
+	}
+
+	if string(rewardedInventory) == "1" {
+		ext[isRewardedInventory] = json.RawMessage(`true`)
+	}
+
+	impExt, err := json.Marshal(ext)
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigital.go]: we need to Marshal and respond with impExt only when `ext` actually changed, i.e. the logic on lines 285-290 should be moved inside the `if string(rewardedInventory)` block.

File: adapters/improvedigital/improvedigital.go [hunk 4]
```diff
@@ -36,6 +48,15 @@ func (a *ImprovedigitalAdapter) MakeRequests(request *openrtb2.BidRequest, reqIn
 }
 
 func (a *ImprovedigitalAdapter) makeRequest(request openrtb2.BidRequest, imp openrtb2.Imp) (*adapters.RequestData, error) {
+	// Handle Rewarded Inventory
+	extImp, err := getImpExtWithRewardedInventory(imp)
+	if err != nil {
+		return nil, err
+	}
+	if len(extImp) > 0 {
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigital.go]: this should be `if extImp != nil {` when you change getImpExtWithRewardedInventory to only respond with `extImp` object when it actually changed

File: adapters/improvedigital/improvedigital.go [hunk 5]
```diff
@@ -114,6 +135,19 @@ func (a *ImprovedigitalAdapter) MakeBids(internalRequest *openrtb2.BidRequest, e
...
 		}
 
+		if bid.Ext != nil {
+			var bidExt BidExt
+			err = json.Unmarshal(bid.Ext, &bidExt)
+			if err != nil {
+				return nil, []error{err}
+			}
+
+			improveDigital := bidExt.ImproveDigital
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigital.go]: please rename `improveDigital` to `bidExtImprovedigital`

File: adapters/improvedigital/improvedigital.go [hunk 6]
```diff
@@ -13,10 +14,21 @@ import (
...
+const BuyingTypeRTB = "rtb"
+const isRewardedInventory = "is_rewarded_inventory"
+
 type ImprovedigitalAdapter struct {
 	endpoint string
 }
 
+// BidExt This struct usage for parse line_item_id and buying_type from bid.ext
+type BidExt struct {
+	ImproveDigital struct {
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigital.go]: please change `D` to `d`, i.e `Improvedigital` as we have it everywhere else

File: adapters/improvedigital/improvedigital.go [hunk 7]
```diff
@@ -114,6 +135,19 @@ func (a *ImprovedigitalAdapter) MakeBids(internalRequest *openrtb2.BidRequest, e
...
 
+		if bid.Ext != nil {
+			var bidExt BidExt
+			err = json.Unmarshal(bid.Ext, &bidExt)
+			if err != nil {
+				return nil, []error{err}
+			}
+
+			improveDigital := bidExt.ImproveDigital
+			if improveDigital.LineItemId != 0 && improveDigital.BuyingType != BuyingTypeRTB {
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigital.go]: BuyingType != BuyingTypeRTB test is not sufficient. check also that it's not an empty string

File: adapters/improvedigital/improvedigitaltest/supplemental/dealid.json [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+                "agency_id": "0"
+              }
+            }
+          },
+          "type": "banner"
+        }
+      ]
+    }
+  ]
+}
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigitaltest/supplemental/dealid.json:594]: please add test cases for when line_item_id is missing or buying_type is missing or empty

**Review by jbartek25 — CHANGES_REQUESTED** [review]
Few minor comments

**Comment by samiul-shanto** [comment]
Master branch updated with `upstream` and rebase `dealid-rewarded-inventory` branch due to security check was failed for `golang.org/x/crypto` package
